src/treelearner/gpu_tree_learner.cpp

```diff
@@ -0,0 +1,1064 @@
+    #endif
+    indices_future_ = boost::compute::copy_async(indices + begin, indices + end, device_data_indices_->begin(), queue_);
+
+    #pragma omp parallel for schedule(static)
+    for (data_size_t i = begin; i < end; ++i) {
+      ordered_hessians_[i - begin] = hessians_[indices[i]];
+    }
+    // copy ordered hessians to the GPU:
+    hessians_future_ = queue_.enqueue_write_buffer_async(device_hessians_, 0, (end - begin) * sizeof(score_t), ptr_pinned_hessians_);
+
+    #pragma omp parallel for schedule(static)
+    for (data_size_t i = begin; i < end; ++i) {
+      ordered_gradients_[i - begin] = gradients_[indices[i]];
+    }
+    // copy ordered gradients to the GPU:
+    gradients_future_ = queue_.enqueue_write_buffer_async(device_gradients_, 0, (end - begin) * sizeof(score_t), ptr_pinned_gradients_);
+
+    #if GPU_DEBUG >= 2
+    Log::Info("gradients/hessians/indiex copied to device with size %d", end - begin);
+    #endif
+  }
+  return SerialTreeLearner::BeforeFindBestSplit(tree, left_leaf, right_leaf);
+}
+
+bool GPUTreeLearner::ConstructGPUHistogramsAsync(
+  const std::vector<int8_t>& is_feature_used,
+  const data_size_t* data_indices, data_size_t num_data,
+  const score_t* gradients, const score_t* hessians,
+  score_t* ordered_gradients, score_t* ordered_hessians) {
+
+  if (num_data <= 0) {
+    return false;
+  }
+
+  
+  // copy data indices if it is not null
+  if (data_indices != nullptr && num_data != num_data_) {
+    indices_future_ = boost::compute::copy_async(data_indices, data_indices + num_data, device_data_indices_->begin(), queue_);
+  }
+  // generate and copy ordered_gradients if gradients is not null
+  if (gradients != nullptr) {
+    if (num_data != num_data_) {
+      #pragma omp parallel for schedule(static)
+      for (data_size_t i = 0; i < num_data; ++i) {
+        ordered_gradients[i] = gradients[data_indices[i]];
+      }
+      gradients_future_ = queue_.enqueue_write_buffer_async(device_gradients_, 0, num_data * sizeof(score_t), ptr_pinned_gradients_);
+    }
+    else {
+      gradients_future_ = queue_.enqueue_write_buffer_async(device_gradients_, 0, num_data * sizeof(score_t), gradients);
+    }
+  }
+  // generate and copy ordered_hessians if hessians is not null
+  if (hessians != nullptr) {
+    if (num_data != num_data_) {
+      #pragma omp parallel for schedule(static)
+      for (data_size_t i = 0; i < num_data; ++i) {
+        ordered_hessians[i] = hessians[data_indices[i]];
+      }
+      hessians_future_ = queue_.enqueue_write_buffer_async(device_hessians_, 0, num_data * sizeof(score_t), ptr_pinned_hessians_);
+    }
+    else {
+      hessians_future_ = queue_.enqueue_write_buffer_async(device_hessians_, 0, num_data * sizeof(score_t), hessians);
+    }
+  }
+  // converted indices in is_feature_used to feature-group indices
+  std::vector<int8_t> is_feature_group_used(num_feature_groups_, 0);
+  for (int i = 0; i < num_features_; ++i) {
```

Comment:
multi-threading here? refer to https://github.com/Microsoft/LightGBM/blob/master/src/io/dataset.cpp#L437-L448